File: apps/xunit-example/CalculatorStoryTest.cs
using ExecutableStories.Xunit;
using Xunit;
using XunitExample;

namespace XunitExample.Tests;

public class CalculatorStoryTest
{
    static CalculatorStoryTest() => Story.Feature(
        "Anyone can do arithmetic without reaching for a calculator app",
        kind: "ability",
        narrative:
            "People doing quick sums in the middle of another task lose their place "
            + "when they have to switch apps. Division throws rather than returning infinity.",
        glossary: new Dictionary<string, string>
        {
            ["operand"] = "One of the two numbers an operation is applied to.",
        });

    // Specified but not built yet: renders as planned in the report.
    [Fact]
    public void Calculator_rejects_non_numeric_input()
    {
        Story.Planned("Calculator rejects non-numeric input");
    }

    [Fact]
    public void Calculator_adds_two_numbers()
    {
        Story.Init("Calculator adds two numbers");
        Story.Given("two numbers 5 and 3");
        var (a, b) = (5, 3);
        Story.When("the numbers are added");
        var result = Calculator.Add(a, b);
        Story.Then("the result is 8");
        Assert.Equal(8, result);
    }

    [Fact]
    public void Calculator_subtracts_two_numbers()
    {
        Story.Init("Calculator subtracts two numbers");
        Story.Given("two numbers 10 and 4");
        var (a, b) = (10, 4);
        Story.When("the second is subtracted from the first");
        var result = Calculator.Subtract(a, b);
        Story.Then("the result is 6");
        Assert.Equal(6, result);
    }

    [Fact]
    public void Calculator_multiplies_two_numbers()
    {
        Story.Init("Calculator multiplies two numbers");
        Story.Given("two numbers 7 and 6");
        var (a, b) = (7, 6);
        Story.Note("This is a note");
        Story.When("the numbers are multiplied");
        var result = Calculator.Multiply(a, b);
        Story.Then("the result is 42");
        Assert.Equal(42, result);
    }

    [Fact]
    public void Calculator_divides_two_numbers()
    {
        Story.Init("Calculator divides two numbers");
        Story.Given("two numbers 20 and 4");
        var (a, b) = (20, 4);
        Story.When("the first is divided by the second");
        var result = Calculator.Divide(a, b);
        Story.Then("the result is 5");
        Assert.Equal(5, result);
    }

    [Fact]
    public void Calculator_throws_on_division_by_zero()
    {
        Story.Init("Calculator throws error on division by zero");
        Story.Note("Division by zero should throw an error");
        Story.Given("a number 10 and zero");
        var (a, b) = (10, 0);
        Story.When("division is attempted");
        Assert.Throws<DivideByZeroException>(() => Calculator.Divide(a, b));
        Story.Then("an error is thrown");
    }
}

File: apps/ruby-example/calculator_story_test.rb
# frozen_string_literal: true

require "minitest/autorun"
require "executable_stories/minitest"
require_relative "calculator"

# Requiring executable_stories/minitest is the whole setup: stories are recorded
# after each test with the status Minitest already worked out.
ExecutableStories.feature(
  kind: "ability",
  title: "Anyone can do arithmetic without reaching for a calculator app",
  narrative: "People doing quick sums in the middle of another task lose their place " \
             "when they have to switch apps. Division raises rather than returning Infinity.",
  glossary: [
    { term: "operand", definition: "One of the two numbers an operation is applied to." }
  ]
)

class CalculatorStoryTest < Minitest::Test
  # Specified but not built yet: renders as planned, stops being planned when
  # someone writes it as a real story.
  def test_rejects_non_numeric_input
    ExecutableStories.planned("Calculator rejects non-numeric input")
  end

  def test_adds_two_numbers
    story = ExecutableStories.init("Calculator adds two numbers")
    story.given("two numbers 5 and 3")
    a = 5
    b = 3
    story.when("the numbers are added")
    result = Calculator.add(a, b)
    story.state({ "a" => a, "b" => b, "result" => result }, label: "Calculator")
    story.then("the result is 8")
    assert_equal 8, result
  end

  def test_subtracts_two_numbers
    story = ExecutableStories.init("Calculator subtracts two numbers")
    story.given("two numbers 10 and 4")
    story.when("the second is subtracted from the first")
    result = Calculator.subtract(10, 4)
    story.then("the result is 6")
    assert_equal 6, result
  end

  def test_multiplies_two_numbers
    story = ExecutableStories.init("Calculator multiplies two numbers")
    story.given("two numbers 7 and 6")
    story.note("Multiplication is repeated addition")
    story.when("the numbers are multiplied")
    result = Calculator.multiply(7, 6)
    story.then("the result is 42")
    assert_equal 42, result
  end

  def test_divides_two_numbers
    story = ExecutableStories.init("Calculator divides two numbers")
    story.given("two numbers 20 and 4")
    story.when("the first is divided by the second")
    result = Calculator.divide(20, 4)
    story.then("the result is 5")
    assert_equal 5, result
  end

  def test_raises_on_division_by_zero
    story = ExecutableStories.init("Calculator raises on division by zero")
    story.note("Division by zero should raise an error")
    story.given("a number 10 and zero")
    story.when("division is attempted")
    story.then("an error is raised")
    assert_raises(ArgumentError) { Calculator.divide(10, 0) }
  end
end
